mqtt.go

```diff
@@ -19,598 +24,613 @@ const (
 
-	switch b.qos {
-	case QOSAtMostOnce:
-		qos = packet.QOSAtMostOnce
-	case QOSAtLeastOnce:
-		qos = packet.QOSAtLeastOnce
-	default:
-		return errors.New("unsupported qos level")
-	}
-
-	p := packet.NewPublishPacket()
-	p.PacketID = mc.getPacketID()
-	p.Message = packet.Message{
-		Topic:   fmt.Sprintf("/devices/%d/bulkData/%s", mc.dc.DeviceID, b.StreamID),
-		QOS:     qos,
-		Payload: b.Blob,
-	}
+	ch := conn.getResponseChan(p)
 
-	mc.outPacket <- p
+	err := conn.writePacket(p)
 
-	logInfo("[MQTT] bulk data delivery packet ID %d", p.PacketID)
-	if b.qos == QOSAtMostOnce {
-		return nil
-	}
+	return ch, err
+}
 
-	logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
+func (conn *mqttConn) runPacketWriter(wg *sync.WaitGroup) {
+	defer func() {
+		logInfo("[MQTT] packet writer is exiting")
+		wg.Done()
+	}()
 
-	select {
-	case <-time.After(deviceEventRetryInterval):
-		msg := fmt.Sprintf("did not receive PUBACK of bulk data delivery for packet ID %d", p.PacketID)
-		logError("[MQTT] %s", msg)
-		mc.reportError(fmt.Errorf("bulk data delivery timeout: %s", msg))
-
-	case ack := <-mc.puback:
-		if ack.PacketID == p.PacketID {
-			logInfo("[MQTT] received PUBACK for packet ID %d", ack.PacketID)
-			return nil
+	// Run until closed
+	for p := range conn.outPktCh {
+		err := conn.writePacket(p)
+		if err != nil {
+			logError("Unable to write out packet %s", p.Type())
+			switch p.Type() {
+			case packet.PINGREQ:
+				conn.pingRespCh <- false
+			case packet.SUBSCRIBE:
+				subPkt := p.(*packet.SubscribePacket)
+				conn.queueAckCh <- MqttQueueResult{subPkt.PacketID, err}
+			case packet.PUBLISH:
+				pubPkt := p.(*packet.PublishPacket)
+				conn.queueAckCh <- MqttQueueResult{pubPkt.PacketID, err}
+			default:
+				// We don't have a packet ID (or did not expect to receive this
+				// packet type and fail in sending it
+				err := fmt.Errorf("Failed to queue packet %s", p.Type())
+				conn.queueAckCh <- MqttQueueResult{0, err}
+			}
 		}
-
-		msg := fmt.Sprintf("sendBulkData() received packet that does not match the Ack packet ID (%d) and the expected packet ID (%d).",
-			ack.PacketID, p.PacketID)
-		logInfo("[MQTT] %s", msg)
-		mc.reportError(fmt.Errorf("bulk data delivery unmatched Ack Packet: %s", msg))
 	}
-
-	return errors.New("bulk data dropped")
 }
 
-func (mc *mqttConn) writeBulkData(b *DeviceSyncedBulkData) error {
-	p := packet.NewPublishPacket()
-	p.PacketID = mc.getPacketID()
-	p.Message = packet.Message{
-		Topic:   fmt.Sprintf("/devices/%d/bulkData/%s", mc.dc.DeviceID, b.StreamID),
-		QOS:     packet.QOSAtLeastOnce,
-		Payload: b.Blob,
-	}
-
-	mc.outPacket <- p
-
-	logInfo("[MQTT] synced bulk data delivery packet ID %d", p.PacketID)
-	logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
+func (conn *mqttConn) runPacketReader(wg *sync.WaitGroup) {
+	defer func() {
+		logInfo("[MQTT] packet reader is exiting")
+		wg.Done()
+		// Close the rest of our channels?
```

Comment:
Yes, we should.
How about using errgroup with context?
https://godoc.org/golang.org/x/sync/errgroup

Reply:
I am closing them in another place now. But, I'll look at the errgroup to see if that's a better solution.